### PR TITLE
Fix inconsistency between layer providers in geometry checker

### LINKED ISSUE #58113: QgsFeatureSource not synchronised with QgsVectorLayer for memory provider
### What is the bug or the crash?

While (still) working on https://github.com/qgis/QGIS-Enhancement-Proposals/issues/236 (porting the geometry checker to the processing toolbox), I stumbled across an incredibly specific bug, which took me days to understand that it was a bug, and days to try to find it...

I need help understanding the bug, and fixing it.

Trying to explain it simply: there is a difference when working with temporary layers VS working with other providers (GPKG in this example).

### Steps to reproduce the issue

The simple test with memory layer:

```python
import tempfile

# Create memory layer and insert a feature
memoryLayer = QgsVectorLayer("Point", "memory", "memory")
f = QgsFeature()
f.setGeometry(QgsGeometry.fromWkt("Point (1 1)"))
assert memoryLayer.startEditing(), "impossible to edit layer"
assert memoryLayer.addFeature(f), "feature not added"
assert memoryLayer.commitChanges(), "changes not committed"

## Test step 1: create a feature source from the layer
featureSource = QgsVectorLayerFeatureSource(memoryLayer)
## Test step 2: change the geometry of the feature
assert memoryLayer.dataProvider().changeGeometryValues({1: QgsGeometry.fromWkt("Point (2 2)")}), "changing geometry on gpkg layer failed"
## Test step 3: verify that the geometry has changed, getting the feature from THE LAYER
assert list(memoryLayer.getFeatures(QgsFeatureRequest(1)))[0].geometry().asWkt() == "Point (2 2)"
## Test step 4: verify that the geometry has changed, getting the feature from THE FEATURE SOURCE
## HERE IS THE ERROR! what do I get here if it is not the changed feature?
assert list(featureSource.getFeatures(QgsFeatureRequest(1)))[0].geometry().asWkt() == "Point (2 2)"
```

The test with memory and GPKG layer showing that GPKG does not fail whereas memory does:

```python
import tempfile

# Create memory layer and insert a feature
memoryLayer = QgsVectorLayer("Point", "memory", "memory")
f = QgsFeature()
f.setGeometry(QgsGeometry.fromWkt("Point (1 1)"))
assert memoryLayer.startEditing(), "impossible to edit layer"
assert memoryLayer.addFeature(f), "feature not added"
assert memoryLayer.commitChanges(), "changes not committed"

# Create GPKG layer from memory layer to have identic layers
with tempfile.NamedTemporaryFile() as tf:
    QgsVectorFileWriter.writeAsVectorFormatV3(memoryLayer, tf.name + ".gpkg", QgsCoordinateTransformContext(), QgsVectorFileWriter.SaveVectorOptions())
    gpkgLayer = QgsVectorLayer(tf.name + ".gpkg")
    assert gpkgLayer.isValid(), "gpkg layer not valid"
    
    ## Test step 1: create a feature source from the layer
    featureSource = QgsVectorLayerFeatureSource(gpkgLayer)
    ## Test step 2: change the geometry of the feature
    assert gpkgLayer.dataProvider().changeGeometryValues({1: QgsGeometry.fromWkt("Point (2 2)")}), "changing geometry on gpkg layer failed"
    ## Test step 3: verify that the geometry has changed, getting the feature from THE LAYER
    assert list(gpkgLayer.getFeatures(QgsFeatureRequest(1)))[0].geometry().asWkt() == "Point (2 2)"
    ## Test step 4: verify that the geometry has changed, getting the feature from THE FEATURE SOURCE
    assert list(featureSource.getFeatures(QgsFeatureRequest(1)))[0].geometry().asWkt() == "Point (2 2)"

## Test step 1: create a feature source from the layer
featureSource = QgsVectorLayerFeatureSource(memoryLayer)
## Test step 2: change the geometry of the feature
assert memoryLayer.dataProvider().changeGeometryValues({1: QgsGeometry.fromWkt("Point (2 2)")}), "changing geometry on gpkg layer failed"
## Test step 3: verify that the geometry has changed, getting the feature from THE LAYER
assert list(memoryLayer.getFeatures(QgsFeatureRequest(1)))[0].geometry().asWkt() == "Point (2 2)"
## Test step 4: verify that the geometry has changed, getting the feature from THE FEATURE SOURCE
## HERE IS THE ERROR! what do I get here if it is not the changed feature?
assert list(featureSource.getFeatures(QgsFeatureRequest(1)))[0].geometry().asWkt() == "Point (2 2)"
```

The CPP test:

```cpp
void TestQgsProcessingFixGeometry::memoryVsGpkg()
{
  QgsVectorLayer memoryLayer = QgsVectorLayer( "Point?crs=2154", "memory", "memory" );

  {
    memoryLayer.startEditing();
    QgsFeature f;
    f.setGeometry( QgsGeometry::fromWkt( "Point (1 1)" ) );
    QVERIFY( memoryLayer.addFeature( f ) );
    QVERIFY( memoryLayer.commitChanges() );
  }

  {
    QgsVectorFileWriter::writeAsVectorFormatV3( &memoryLayer, "/home/jvolpes/Desktop/test.gpkg", QgsCoordinateTransformContext(), QgsVectorFileWriter::SaveVectorOptions() );
    QgsVectorLayer gpkgLayer = QgsVectorLayer( "/home/jvolpes/Desktop/test.gpkg|layername=test" );
    QgsVectorLayerFeatureSource mFeatureSource( &gpkgLayer );

    QgsGeometryMap geometryMap;
    geometryMap.insert( 1, QgsGeometry::fromWkt( "Point (2 2)" ) );
    QVERIFY( gpkgLayer.dataProvider()->changeGeometryValues( geometryMap ) );

    QgsFeature nf;
    gpkgLayer.getFeatures( QgsFeatureRequest( 1 ) ).nextFeature( nf );
    QCOMPARE( nf.geometry().asWkt(), "Point (2 2)" );

    mFeatureSource.getFeatures( QgsFeatureRequest( 1 ) ).nextFeature( nf );
    QCOMPARE( nf.geometry().asWkt(), "Point (2 2)" );
  }

  {
    QgsVectorLayerFeatureSource mFeatureSource( &memoryLayer );

    QgsGeometryMap geometryMap;
    geometryMap.insert( 1, QgsGeometry::fromWkt( "Point (2 2)" ) );
    QVERIFY( memoryLayer.dataProvider()->changeGeometryValues( geometryMap ) );

    QgsFeature nf;
    memoryLayer.getFeatures( QgsFeatureRequest( 1 ) ).nextFeature( nf );
    QCOMPARE( nf.geometry().asWkt(), "Point (2 2)" );

    mFeatureSource.getFeatures( QgsFeatureRequest( 1 ) ).nextFeature( nf );
    QCOMPARE( nf.geometry().asWkt(), "Point (2 2)" );
  }
}
```

### Versions

master and LTR 3.34.8

### Supported QGIS version

- [ ] I'm running a supported QGIS version according to [the roadmap](https://www.qgis.org/en/site/getinvolved/development/roadmap.html#release-schedule).

### New profile

- [ ] I tried with a new [QGIS profile](https://docs.qgis.org/latest/en/docs/user_manual/introduction/qgis_configuration.html#working-with-user-profiles)

### Additional context

Any help is welcome, I need to know if it's supposed to work as it works with memory or ogr provider, and advices on how to fix it, or point code locations... 

Thank you.

### PR DESCRIPTION
**Pull request by Djedouas**
Geometry checker does not work properly with memory layers, so we change how the cache in the QgsFeaturePool object is managed, to refresh it more often.

Fixes #58113

### CODE REVIEW

File: src/analysis/vector/geometry_checker/qgsfeaturepool.cpp [hunk 1]
```diff
@@ -58,6 +58,7 @@ bool QgsFeaturePool::getFeature( QgsFeatureId id, QgsFeature &feature )
   {
     // Feature not in cache, retrieve from layer
     // TODO: avoid always querying all attributes (attribute values are needed when merging by attribute)
+    mFeatureSource = std::make_unique<QgsVectorLayerFeatureSource>( mLayer );
```
Comment (nyalldawson) [inline src/analysis/vector/geometry_checker/qgsfeaturepool.cpp:61]: Unfortunately this fix isn't thread safe -- you can't create a feature source from a different thread vs what the layer lives in

**Review by nyalldawson — CHANGES_REQUESTED** [review]

**Comment by Djedouas** [comment]
@lbartoletti

**Comment by nyalldawson** [comment]
I have to look further into this one, but I believe there's thread safety issues in this change. Please don't merge yet.

File: src/analysis/vector/geometry_checker/qgsfeaturepool.cpp [hunk 2]
```diff
@@ -33,6 +33,7 @@ QgsFeaturePool::QgsFeaturePool( QgsVectorLayer *layer )
   , mGeometryType( layer->geometryType() )
   , mFeatureSource( std::make_unique<QgsVectorLayerFeatureSource>( layer ) )
   , mLayerName( layer->name() )
+  , mIsMemoryLayer( layer->dataProvider()->name() == QStringLiteral( "memory" ) )
```
Comment (nyalldawson) [inline src/analysis/vector/geometry_checker/qgsfeaturepool.cpp:36]: Slightly safer, won't crash if the layer is invalid:

```suggestion
  , mIsMemoryLayer( layer->providerType() == QStringLiteral( "memory" ) )
```

File: src/analysis/vector/geometry_checker/qgsfeaturepool.cpp [hunk 3]
```diff
@@ -47,6 +48,17 @@ bool QgsFeaturePool::getFeature( QgsFeatureId id, QgsFeature &feature )
   //
   // https://bugreports.qt.io/browse/QTBUG-19794
 
+  // a memory layer is already a cache itself. retrieve feature from the feature source.
+  if ( mIsMemoryLayer )
+  {
+    if ( !mLayer->getFeatures( QgsFeatureRequest( id ) ).nextFeature( feature ) )
```
Comment (nyalldawson) [inline src/analysis/vector/geometry_checker/qgsfeaturepool.cpp:54]: This isn't thread safe -- eg. in the geometry checker plugin the feature pool is used from a background thread after constructing the source (which IS a thread safe approach, the feature source was created on the main thread), but we can't just swap this out to access the layer here because accessing the **layer** from a different thread is not thread safe.

**Comment by nyalldawson** [comment]
@Djedouas see some notes here  -- https://github.com/qgis/QGIS/pull/58132

File: src/analysis/vector/geometry_checker/qgsfeaturepool.cpp (hunk 3 above)
Comment (Djedouas) [reply src/analysis/vector/geometry_checker/qgsfeaturepool.cpp:54]: Right, I see the difference between "being in another thread and accessing the feature source created on the main thread" and "being in another thread and accessing the layer created on the main thread".

It is complicated, today the geometry checker only works because updating the feature source **does** indeed update the underlying data (we discussed this in #58113)... which is not the case with a memory layer...

In fact, the question is "how do I get a feature that is up-to-date for a memory layer that lies in another thread?"

What are your thoughts about:
- option 1: encapsulate this not-thread-safe statement in a lambda with `QgsThreadingUtils::runOnMainThread()`
- option 2: have a check that the layer is in the correct thread and return `false` or handle an exception?
- option 3: make `getFeatures()` for a memory layer thread-safe, with mutex etc. ?

Comment (Djedouas) [reply src/analysis/vector/geometry_checker/qgsfeaturepool.cpp:54]: @nyalldawson this issue is very important, because the PR for exposing all geometry checker algorithms to the processing toolbox depends on it (the code is ready, I could open the PR, but it relies on this commit at the moment).

Comment (nyalldawson) [reply src/analysis/vector/geometry_checker/qgsfeaturepool.cpp:54]: @Djedouas 

Honestly, I don't particularly like any of those 3 options. Especially not 3. Don't do 3. Run from that one :rofl: 

I think trying to patch around this in the QgsFeaturePool class like this might be the wrong approach. My understanding of the issue is that the geometry checker is changing features in the layer, and then the feature cached in the pool is out of date with that. If so, then a couple of things jump to mind:

1. Since the geometry checker has knowledge of the feature pool already, what about adding an explicit "set updated feature" method to the pool? There could be a QHash< QgsFeatureId, QgsFeature > to handle this, so when "setUpdatedFeature" is called then the feature is added to this hash. Then getFeature would first check if the feature is the updated feature hash, and if so, it just returns that feature instead of querying the feature source. Admittedly it's not the most elegant solution, but it would be a pragmatic fix for the problem which doesn't have any risk of deadlocks (as runOnMainThread would have), and doesn't involve messing with any of the fundamental, low-level threading design around the feature sources/vector layers (as adding the mutex to the memory provider would do).
2. I think there's potentially a deeper issue at play here too :scream: . Are you updating features in the vector layer from a background thread? (eg in a processing algorithm's `processAlgorithm` function). If so, that's not thread safe, and should NOT be done. You instead need to queue up the changes for the layer, and then apply them ONLY in the postProcessAlgorithm step, which runs in the main thread after processAlgorithm finishes.

Comment (Djedouas) [reply src/analysis/vector/geometry_checker/qgsfeaturepool.cpp]: @nyalldawson thanks, see my thoughts below

About point 1.

Your point made me think of another way to fix the specific memory layer issue. You're right, we have a cache that is out of date with the real geometry, and the geometry checker gets the feature from the cache when fixing geometry checks. But I thought that, other than creating yet another sub-cache with up-do-date features, I could take a look at the method refreshing the cache.

So here is a new commit / new way to fix the problem: in `refreshCache` (which was using `getFeature`, and falling in the no-cache branch there, and getting the out-of-date geometry from the feature source) we can directly set the up-to-date feature and its geometry.

About point 2.

In all incoming processing algorithms that will use the geometry checker and feature pool logic, a copy of the input layer is created within the `processAlgorithm` function, I don't use the multi-threading code of the current geometry checker plugin, and everything checking/modifying the layer is done in the processing thread, where the layer is from. We can discuss this further on in the dedicated PR I will submit.

File: src/analysis/vector/geometry_checker/qgsfeaturepool.h [hunk 4]
```diff
@@ -174,7 +174,7 @@ class ANALYSIS_EXPORT QgsFeaturePool : public QgsFeatureSink SIP_ABSTRACT
      * \note This method can safely be called from a different thread vs the object's creation thread or
      * the original layer's thread.
      */
-    void refreshCache( const QgsFeature &feature );
+    void refreshCache( QgsFeature feature );
```
Comment (nyalldawson) [inline src/analysis/vector/geometry_checker/qgsfeaturepool.h:177]: Why this change?

File: src/analysis/vector/geometry_checker/qgsfeaturepool.cpp [hunk 5]
```diff
@@ -131,15 +131,13 @@ void QgsFeaturePool::insertFeature( const QgsFeature &feature, bool skipLock )
   mIndex.addFeature( indexFeature );
 }
 
-void QgsFeaturePool::refreshCache( const QgsFeature &feature )
+void QgsFeaturePool::refreshCache( QgsFeature feature )
 {
   QgsReadWriteLocker locker( mCacheLock, QgsReadWriteLocker::Write );
-  mFeatureCache.remove( feature.id() );
+  mFeatureCache.insert( feature.id(), new QgsFeature( feature ) );
   mIndex.deleteFeature( feature );
```
Comment (nyalldawson) [inline src/analysis/vector/geometry_checker/qgsfeaturepool.cpp:138]: Looking at this, I'm actually unsure how this ever worked without crashes :upside_down_face: 

QgsSpatialIndex requires that the ORIGINAL feature added to the index is passed. It's not operating on the feature ID alone, but rather the combination of the feature's bounding box + feature ID. So here we ask the spatial index to remove a feature which has a different bounding box from the spatial index, which will corrupt the index! :scream: 

::removeFeature works OK, because it's retrieving the original feature from the feature ID and then removing that from the index. We need to do similar here -- first retrieve the original feature, remove from the index, and THEN update the cache/spatial index.

But that's still fraught with potential traps -- eg the subclasses are updating the original feature first before calling refreshCache, so if the feature has been evicted from mFeatureCache already then we'll end up retrieving the UPDATED feature from the call to getFeature and the spatial index will still be corrupted.

Maybe we'll need to add a QHash< QgsFeatureId, QgsRectangle > mOriginalFeatureBounds here and use that with `QgsSpatialIndex::deleteFeature( QgsFeatureId id, const QgsRectangle &bounds )` instead...

**Review by nyalldawson — CHANGES_REQUESTED** [review]

File: src/analysis/vector/geometry_checker/qgsfeaturepool.h (hunk 4 above)
Comment (Djedouas) [reply src/analysis/vector/geometry_checker/qgsfeaturepool.h:177]: `QgsSpatialIndex::addFeature` doesn't have `const` qualifier on its input feature, so I made this change to pass by copy.

File: src/analysis/vector/geometry_checker/qgsfeaturepool.cpp (hunk 5 above)
Comment (Djedouas) [reply src/analysis/vector/geometry_checker/qgsfeaturepool.cpp]: All right... I think we are getting closer to the solution... 😆 

> We need to do similar here -- first retrieve the original feature, remove from the index, and THEN update the cache/spatial index.

I see 😄 

Here is a new commit, I think we can do without an additional class member, as we can manage to get the original feature geometry each time we need to refresh the index and give it to `::refreshCache`

Also adapted `QgsVectorLayerFeaturePool::onGeometryChanged` and its test, because the intent of was clearly to keep the cache up to date when an external geometry change occurs.

File: src/analysis/vector/geometry_checker/qgsvectorlayerfeaturepool.cpp [hunk 6]
```diff
@@ -108,6 +108,9 @@ bool QgsVectorLayerFeaturePool::addFeatures( QgsFeatureList &features, QgsFeatur
 
 void QgsVectorLayerFeaturePool::updateFeature( QgsFeature &feature )
 {
+  QgsFeature origFeature;
+  getFeature( feature.id(), origFeature );
```
Comment (nyalldawson) [inline src/analysis/vector/geometry_checker/qgsvectorlayerfeaturepool.cpp:112]: I may be getting confused here, but I think there's still a risk of index corruption in this scenario:

1. Feature A is retrieved (version "1"), it is cached in the pool. Bounds are used in the spatial index.
2. More features are fetched, cache fills up, feature A is evicted from cache
3. Some external caller changes the geometry on feature A in the vector layer or data provider, now there's version 2 of feature A.
4. A call to updateFeature for feature A is made. getFeature won't find it in the pool's cache (it was evicted), so it's retrieved from the layer. We get feature A version 2.
5. Version 2 is passed to refreshCache, and the index is corrupted because we try to remove different bounds (version 2, whereas the cache is using v1)

I don't see any alternative but keeping a local map/hash of the feature bounds which were used in the index.

Comment (Djedouas) [reply src/analysis/vector/geometry_checker/qgsvectorlayerfeaturepool.cpp:112]: You're right, I did not expect that the spatial index could evict a feature... It is an index, not strictly a cache, isn't it?

Comment (nyalldawson) [reply src/analysis/vector/geometry_checker/qgsvectorlayerfeaturepool.cpp:112]: It won't get evicted from the spatial index, but will from mFeatureCache.

Comment (Djedouas) [reply src/analysis/vector/geometry_checker/qgsvectorlayerfeaturepool.cpp:112]: Of course 🙈   obviously.

Okay, what about "synchronizing" the cache and the index?

In fact, in the end of step 4. inside the index, we have both feature A version 2 and feature A version 1.

We could fix this: when getFeature doesn't find a cached feature, we can proactively delete any pre-existing corresponding feature in the index (see new commit).

If the feature was not present in the index, it does nearly nothing, otherwise it removes the possibly out-of-date feature from the index, to re-add the possibly same or up-to-date feature.

File: src/analysis/vector/geometry_checker/qgsfeaturepool.cpp [hunk 7]
```diff
@@ -66,6 +66,9 @@ bool QgsFeaturePool::getFeature( QgsFeatureId id, QgsFeature &feature )
     }
     locker.changeMode( QgsReadWriteLocker::Write );
     mFeatureCache.insert( id, new QgsFeature( feature ) );
+
+    //cleanup the index, in case the feature was evicted from the cache but is still in the index
+    mIndex.deleteFeature( id );
```
Comment (nyalldawson) [inline src/analysis/vector/geometry_checker/qgsfeaturepool.cpp:71]: This doesn't do what it appears to do :scream: 

- there's no QgsSpatialIndex::deleteFeature( QgsFeatureId id ) method
- The feature ID is triggering an implicit construction of QgsFeature via the `QgsFeature( QgsFeatureID id )` constructor (which should possibly be marked explicit..., but don't touch that here!). So then QgsSpatialIndex::deleteFeature is being called with a QgsFeature which has NO geometry
- Follow the logic in QgsSpatialIndex::deleteFeature -- it requires the passed feature geometry to exist and match the geometry of the feature currently present in the index.

I think you better add some tests triggering the cache eviction logic bug first before implementing a fix :stuck_out_tongue_closed_eyes:

File: src/analysis/vector/geometry_checker/qgsfeaturepool.cpp [hunk 8]
```diff
@@ -66,7 +66,21 @@ bool QgsFeaturePool::getFeature( QgsFeatureId id, QgsFeature &feature )
     }
     locker.changeMode( QgsReadWriteLocker::Write );
     mFeatureCache.insert( id, new QgsFeature( feature ) );
-    mIndex.addFeature( feature );
+
+    QgsGeometry indexGeom = mIndex.geometry( id );
```
Comment (nyalldawson) [inline src/analysis/vector/geometry_checker/qgsfeaturepool.cpp:70]: The index isn't created with the FlagStoreFeatureGeometries flag, so this will always be null.

Can you please add some tests before trying more solutions? Right now we're just fumbling in the dark :rofl:

File: src/analysis/vector/geometry_checker/qgsfeaturepool.cpp (hunk 7 above)
Comment (nyalldawson) [reply src/analysis/vector/geometry_checker/qgsfeaturepool.cpp]: I'm still -1 to introducing further known issues into this class, when we know upfront that their around and will cause someone to lose a lot of time re-discovering this and fixing in future.

Why not just go with my original proposal of adding the bounding box hash to the feature pool? That's a couple of lines of code to add, and will fix the issue properly.

File: src/analysis/vector/geometry_checker/qgsfeaturepool.cpp (hunk 8 above)
Comment (Djedouas) [reply src/analysis/vector/geometry_checker/qgsfeaturepool.cpp:70]: > The index isn't created with the FlagStoreFeatureGeometries flag, so this will always be null.

The index creation here https://github.com/qgis/QGIS/blob/master/src/analysis/vector/geometry_checker/qgsfeaturepool.cpp#L82 has default flags which seems ON for FlagStoreFeatureGeometries https://github.com/qgis/QGIS/blob/master/src/core/qgsspatialindex.h#L77

Comment (nyalldawson) [reply src/analysis/vector/geometry_checker/qgsfeaturepool.cpp:70]: That's not how qflags work 😂

Comment (Djedouas) [reply src/analysis/vector/geometry_checker/qgsfeaturepool.cpp:70]: ...yes I was expecting that 😆   writing a test at the moment... ^^

File: src/analysis/vector/geometry_checker/qgsfeaturepool.cpp (hunk 7 above)
Comment (Djedouas) [reply src/analysis/vector/geometry_checker/qgsfeaturepool.cpp]: > Why not just go with my original proposal of adding the bounding box hash to the feature pool? That's a couple of lines of code to add, and will fix the issue properly.

Yes I was hoping we could use the spatial index ::geometry() method like such a hash, it seems to behave exactly like this (with FlagStoreFeatureGeometries set 😆 )

Comment (Djedouas) [reply src/analysis/vector/geometry_checker/qgsfeaturepool.cpp]: Oh but it would store whole geometry data rather than only small bounding boxes... 

Ok let's go with the hash, I have the impression to triple the data with feature cache, index, bounding boxes, but I agree that it is necessary in the end (and I have learned a lot 👍🏼  thanks)

**Comment by Djedouas** [comment]
I prefered to create a new PR to facilitate discussion, as I followed @nyalldawson advice to use a `QHash<QgsFeatureId, QgsFeature> mUpdatedFeatures` hash in the class.
